python/ql/src/Security/CWE-327/Ssl.qll

```diff
@@ -0,0 +1,208 @@
+      notFlag.getOperand() = flag and
+      flag = API::moduleImport("ssl").getMember("OP_NO_" + restriction).getAUse().asExpr() and
+      (
+        aa.getValue() = notFlag
+        or
+        impliesBitSet(aa.getValue(), notFlag, true, true)
+      )
+    )
+  }
+
+  override DataFlow::CfgNode getContext() { result = this }
+
+  override ProtocolVersion getUnrestriction() { result = restriction }
+}
+
+/**
+ * Holds if
+ *   for every bit, _b_:
+ *     `wholeHasBitSet` represents that _b_ is set in `whole`
+ *     implies
+ *     `partHasBitSet` represents that _b_ is set in `part`
+ *
+ * As an example take `whole` = `part1 & part2`. Then
+ * `impliesBitSet(whole, part1, true, true)` holds
+ * because for any bit in `whole`, if that bit is set it must also be set in `part1`.
+ *
+ * Similarly for `whole` = `part1 | part2`. Here
+ * `impliesBitSet(whole, part1, false, false)` holds
+ * because for any bit in `whole`, if that bit is not set, it cannot be set in `part1`.
+ */
+predicate impliesBitSet(BinaryExpr whole, Expr part, boolean partHasBitSet, boolean wholeHasBitSet) {
+  whole.getOp() instanceof BitAnd and
+  (
+    wholeHasBitSet = true and partHasBitSet = true and part in [whole.getLeft(), whole.getRight()]
+    or
+    wholeHasBitSet = true and
+    impliesBitSet([whole.getLeft(), whole.getRight()], part, partHasBitSet, wholeHasBitSet)
+  )
+  or
+  whole.getOp() instanceof BitOr and
+  (
+    wholeHasBitSet = false and partHasBitSet = false and part in [whole.getLeft(), whole.getRight()]
+    or
+    wholeHasBitSet = false and
+    impliesBitSet([whole.getLeft(), whole.getRight()], part, partHasBitSet, wholeHasBitSet)
+  )
+}
+
+class ContextSetVersion extends ProtocolRestriction, ProtocolUnrestriction {
+  ProtocolVersion restriction;
+
+  ContextSetVersion() {
+    exists(DataFlow::AttrWrite aw |
+      aw.getObject().asCfgNode() = node and
+      aw.getAttributeName() = "minimum_version" and
+      aw.getValue() =
+        API::moduleImport("ssl").getMember("TLSVersion").getMember(restriction).getAUse()
+    )
+  }
+
+  override DataFlow::CfgNode getContext() { result = this }
+
+  override ProtocolVersion getRestriction() { result.lessThan(restriction) }
+
+  override ProtocolVersion getUnrestriction() {
+    restriction = result or restriction.lessThan(result)
+  }
+}
+
+class UnspecificSSLContextCreation extends SSLContextCreation, UnspecificContextCreation {
+  UnspecificSSLContextCreation() { library = "ssl" }
+
+  override ProtocolVersion getUnrestriction() {
+    // Case: A protocol argument is present.
+    result = UnspecificContextCreation.super.getUnrestriction() and
+    // These are turned off by default
+    // see https://docs.python.org/3/library/ssl.html#ssl-contexts
+    not result in ["SSLv2", "SSLv3"]
+    or
+    // Case: No protocol arguemnt is present.
+    not exists(this.getProtocol()) and
+    // The default argument is TLS and the SSL versions are turned off by default.
```

Comment:
```suggestion
    // The default argument is TLS and the SSL versions are turned off by default since Python 3.6
    // see https://docs.python.org/3.6/library/ssl.html#ssl.SSLContext
```